File: CourseSchedule_LC207.java
/*
T: O(V+E)
S: O(V+E) : hashmap(V+E) + indegrees(V) + queue(V) -> E+3V => O(V+E)
where V = no of courses
E = no of edges -> size of prereq array
*/

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        //build in-degree array to find independent nodes
        int[] indegrees = new int[numCourses];
        
        //build a map of courses: prereq -> next course
        Map<Integer, List<Integer>> map = new HashMap<>();
        
        for(int[] prereq: prerequisites){
            int post = prereq[1];
            int pre = prereq[0];
            map.computeIfAbsent(post, i -> new ArrayList<>()).add(pre); 
            
            //update indegrees array
            indegrees[pre]++;
 
        }
        
        int count = 0; // no of courses finished
        Queue<Integer> q = new LinkedList<>();
        //add all independent nodes to the queue
        for(int i = 0; i < numCourses; i++){
            if(indegrees[i] == 0)
                q.add(i);
        }
        
        //return if no independent nodes i.e. q is empty
        if(q.isEmpty()) return false;
        
        //now take the independent nodes out and reduce the dependency of post courses
        while(!q.isEmpty()){
            int pre = q.poll();
            count++; //finished the course
            //update indegrees
            List<Integer> courses = map.get(pre);
            if(courses != null){
                for(int c: courses){
                indegrees[c]--;
                if(indegrees[c] == 0)
                    q.add(c);
                }   
            }
        }
        
        return count == numCourses ? true : false;
    }
}

File: course-schedule_LC_207.py
#TC:O(M+N) SC:(M+N)
#Graph problem

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        array=[]
        dic={}
        queue=[]
        count = 0
        
        for i in range(numCourses):
            array.append(0)

        for i in prerequisites:
            array[i[0]]+=1
            if i[1] in dic:
                dic[i[1]].append(i[0])
            else: 
                dic[i[1]]=[i[0]]      

        for i,val in enumerate(array):
            if val==0:
                count+=1
                queue.append(i)

        if queue==None:
            return False

        while queue!=[]:
            curr=queue.pop(0)
            if curr not in dic:
                continue
            else:
                for j in dic[curr]:
                    array[j]-=1
                    if array[j]==0:
                        count+=1
                        queue.append(j)

        return count==numCourses
